File: modules/openlmis-web/src/test/java/org/openlmis/web/controller/SupplyLineControllerTest.java
/*
 * This program is part of the OpenLMIS logistics management information system platform software.
 * Copyright © 2013 VillageReach
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *  
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.
 * You should have received a copy of the GNU Affero General Public License along with this program.  If not, see http://www.gnu.org/licenses.  For additional information contact info@OpenLMIS.org. 
 */

package org.openlmis.web.controller;

import org.hamcrest.CoreMatchers;
import org.junit.Before;
import org.junit.Test;
import org.junit.experimental.categories.Category;
import org.junit.runner.RunWith;
import org.junit.runners.BlockJUnit4ClassRunner;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.openlmis.core.domain.Pagination;
import org.openlmis.core.domain.SupplyLine;
import org.openlmis.core.exception.DataException;
import org.openlmis.core.service.MessageService;
import org.openlmis.core.service.SupplyLineService;
import org.openlmis.db.categories.UnitTests;
import org.openlmis.core.web.OpenLmisResponse;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;
import org.powermock.modules.junit4.PowerMockRunnerDelegate;
import org.springframework.http.ResponseEntity;
import org.springframework.mock.web.MockHttpServletRequest;

import java.util.List;

import static java.util.Arrays.asList;
import static org.hamcrest.core.Is.is;
import static org.junit.Assert.assertThat;
import static org.mockito.Matchers.anyString;
import static org.mockito.Mockito.*;
import static org.mockito.MockitoAnnotations.initMocks;
import static org.openlmis.web.controller.SupplyLineController.PAGINATION;
import static org.openlmis.web.controller.SupplyLineController.SUPPLY_LINES;
import static org.powermock.api.mockito.PowerMockito.when;
import static org.powermock.api.mockito.PowerMockito.whenNew;

@Category(UnitTests.class)
@RunWith(PowerMockRunner.class)
@PowerMockRunnerDelegate(BlockJUnit4ClassRunner.class)
@PrepareForTest(SupplyLineController.class)
public class SupplyLineControllerTest {

  @Mock
  SupplyLineService service;

  @Mock
  MessageService messageService;

  @InjectMocks
  SupplyLineController controller;

  public static final String USER_ID = "USER_ID";

  private MockHttpServletRequest request = new MockHttpServletRequest();

  @Before
  public void setUp() {
    initMocks(this);
    request = new MockHttpServletRequest();
    request.getSession().setAttribute("USER_ID", 1L);
  }

  @Test
  public void shouldSearchSupplyLines() throws Exception {
    String searchParam = "supply";
    String column = "name";
    Integer page = 2;
    String limit = "10";
    List<SupplyLine> supplyLines = asList(new SupplyLine());

    Pagination pagination = new Pagination(2, 3);
    whenNew(Pagination.class).withArguments(page, Integer.parseInt(limit)).thenReturn(pagination);
    when(service.search(searchParam, column, pagination)).thenReturn(supplyLines);
    when(service.getTotalSearchResultCount(searchParam, column)).thenReturn(3);

    ResponseEntity<OpenLmisResponse> response = controller.search(searchParam, column, page, limit);

    assertThat((List<SupplyLine>) response.getBody().getData().get(SUPPLY_LINES), is(supplyLines));
    assertThat((Pagination) response.getBody().getData().get(PAGINATION), is(pagination));
  }

  @Test
  public void shouldInsertSupplyLineSuccessfully() {
    SupplyLine supplyLine = new SupplyLine(1L);
    Mockito.when(messageService.message("message.supply.line.created.success")).thenReturn("success");

    ResponseEntity<OpenLmisResponse> responseEntity = controller.insert(supplyLine, request);

    verify(service).save(supplyLine);
    assertThat((Long) responseEntity.getBody().getData().get("supplyLineId"), CoreMatchers.is(supplyLine.getId()));
    assertThat(responseEntity.getBody().getSuccessMsg(), CoreMatchers.is("success"));
    assertThat(supplyLine.getCreatedBy(), CoreMatchers.is(1L));
    assertThat(supplyLine.getModifiedBy(), CoreMatchers.is(1L));
  }

  @Test
  public void shouldReturnErrorMessageWhenExceptionOccursOnSupplyLineInsert() {
    SupplyLine supplyLine = new SupplyLine();
    doThrow(new DataException("error")).when(service).save(supplyLine);

    ResponseEntity<OpenLmisResponse> responseEntity = controller.insert(supplyLine, request);

    verify(service).save(supplyLine);
    verify(messageService, never()).message(anyString(), anyString());
    assertThat(responseEntity.getBody().getErrorMsg(), CoreMatchers.is("error"));
    assertThat(supplyLine.getCreatedBy(), CoreMatchers.is(1L));
    assertThat(supplyLine.getModifiedBy(), CoreMatchers.is(1L));
  }

  @Test
  public void shouldUpdateSupplyLineSuccessfully() {
    SupplyLine supplyLine = new SupplyLine(1L);
    Mockito.when(messageService.message("message.supply.line.updated.success")).thenReturn("success");

    ResponseEntity<OpenLmisResponse> responseEntity = controller.update(supplyLine, supplyLine.getId(), request);

    verify(service).save(supplyLine);
    assertThat((Long) responseEntity.getBody().getData().get("supplyLineId"), CoreMatchers.is(supplyLine.getId()));
    assertThat(responseEntity.getBody().getSuccessMsg(), CoreMatchers.is("success"));
    assertThat(supplyLine.getModifiedBy(), CoreMatchers.is(1L));
  }

  @Test
  public void shouldReturnErrorMessageWhenExceptionOccursOnSupplyLineUpdate() {
    SupplyLine supplyLine = new SupplyLine(1L);
    doThrow(new DataException("error")).when(service).save(supplyLine);

    ResponseEntity<OpenLmisResponse> responseEntity = controller.update(supplyLine, 1L, request);

    verify(service).save(supplyLine);
    verify(messageService, never()).message(anyString(), anyString());
    assertThat(responseEntity.getBody().getErrorMsg(), CoreMatchers.is("error"));
    assertThat(supplyLine.getModifiedBy(), CoreMatchers.is(1L));
  }

  @Test
  public void shouldGetSupervisoryNodeById() {
    Long id = 1L;
    SupplyLine expectedSupplyLine = new SupplyLine();
    Mockito.when(service.getById(id)).thenReturn(expectedSupplyLine);

    SupplyLine supplyLine = controller.getById(id);

    verify(service).getById(id);
    assertThat(supplyLine, CoreMatchers.is(expectedSupplyLine));
  }

}

File: modules/openlmis-web/src/test/javascript/specs/supply-line/controller/supply-line-controller-test.js
/*
 * This program is part of the OpenLMIS logistics management information system platform software.
 * Copyright © 2013 VillageReach
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.
 * You should have received a copy of the GNU Affero General Public License along with this program.  If not, see http://www.gnu.org/licenses.  For additional information contact info@OpenLMIS.org.
 */

describe("Supply Line Controller", function () {

  var scope, ctrl, supplyLine, $httpBackend, location, programs;
  beforeEach(module('openlmis'));

  beforeEach(inject(function ($rootScope, _$httpBackend_, $controller, $location) {
    scope = $rootScope.$new();
    $httpBackend = _$httpBackend_;
    location = $location;
    scope.query = "Nod";
    supplyLine = {"program": {"name": "TB"}, "supplyingFacility": {"name": "supplying"}};
    programs = [
      {"name": "TB", "id": 1},
      {"name": "MALARIA", "id": 2}
    ];
    ctrl = $controller('SupplyLineController', {$scope: scope, supplyLine: supplyLine, programs: programs});
  }));

  it('should set supply Line in scope', function () {
    expect(scope.supplyLine).toBe(supplyLine);
  });

  it('should get all top level nodes in scope if query length is greater than 3 in case of add', function () {
    scope.query = "Nod";
    var node1 = {"id": 1, "code": "N1", "name": "Node 1"};
    var node2 = {"id": 2, "code": "N2", "name": "Node 2"};
    var node3 = {"id": 3, "code": "N3", "name": "Node 3"};
    var response = {"supervisoryNodeList": [node1, node2, node3]};

    $httpBackend.when('GET', '/topLevelSupervisoryNodes.json?searchParam=' + scope.query).respond(response);
    scope.showTopLevelNodeResults();
    $httpBackend.flush();

    expect(scope.topLevelNodes).toEqual([node1, node2, node3]);
    expect(scope.filteredNodes).toEqual([node1, node2, node3]);
    expect(scope.previousQuery).toEqual("Nod");
    expect(scope.nodeResultCount).toEqual(3);
  });

  it('should do client side filtering when previous query and current query are same', function () {
    spyOn($httpBackend, "expectGET");
    scope.query = "Node";
    scope.previousQuery = "Nod";
    var node2 = {"id": 2, "code": "N2", "name": "Node 2"};
    var node3 = {"id": 3, "code": "N3", "name": "Node 3"};
    var node4 = {"id": 3, "code": "N3", "name": "Nodde 4"};
    scope.topLevelNodes = [node2, node3, node4];

    scope.showTopLevelNodeResults();

    expect($httpBackend.expectGET).not.toHaveBeenCalledWith('/topLevelSupervisoryNodes.json?searchParam=' + scope.query);
    expect(scope.filteredNodes).toEqual([node2, node3]);
    expect(scope.previousQuery).toEqual("Nod");
    expect(scope.nodeResultCount).toEqual(2);
  });

  it('should not search results if query is undefined', function () {
    spyOn($httpBackend, "expectGET");
    scope.query = undefined;

    scope.showTopLevelNodeResults();

    expect($httpBackend.expectGET).not.toHaveBeenCalledWith('/topLevelSupervisoryNodes.json?searchParam=' + scope.query);
  });

  it('should not search results if query length is less than 3', function () {
    spyOn($httpBackend, "expectGET");
    scope.query = "No";

    scope.showTopLevelNodeResults();

    expect($httpBackend.expectGET).not.toHaveBeenCalledWith('/topLevelSupervisoryNodes.json?searchParam=' + scope.query);
  });

  it('should select supervisory node and clear scope fields', function () {
    var node = {code: "SN2", name: "Node 2"};
    scope.setSupervisoryNode(node);
    expect(scope.supplyLine.supervisoryNode).toBe(node);
    expect(scope.query).toBeUndefined();
    expect(scope.nodeResultCount).toBeUndefined();
    expect(scope.filteredFacilities).toBeUndefined();
    expect(scope.previousQuery).toBeUndefined();
  });

  it('should delete supervisory node', function () {
    scope.supplyLine.parent = {"Code": "N1", "Name": "Node 1"};

    scope.deleteSupervisoryNode();

    expect(scope.supplyLine.supervisoryNode).toBeUndefined();
  });

  it('should take to search page on cancel', function () {
    scope.cancel();
    expect(scope.$parent.message).toEqual("");
    expect(scope.$parent.supplyLineId).toBeUndefined();
    expect(location.path()).toEqual('/#/search');
  });

  it('should not save supply Line if invalid', function () {
    scope.supplyLineForm = {"$error": {"required": true}};

    scope.save();

    expect(scope.error).toEqual("form.error");
    expect(scope.showError).toBeTruthy();
  });

  it('should not save supply Line if supplying facility missing', function () {
    scope.supplyLineForm = {"$error": {"required": false}};
    scope.supplyLine = {supplyingFacility: undefined};

    scope.save();

    expect(scope.error).toEqual("form.error");
    expect(scope.showError).toBeTruthy();
  });

  it('should not save supply Line if supervisory node missing', function () {
    scope.supplyLineForm = {"$error": {"required": false}};
    scope.supplyLine = {supervisoryNode: undefined};

    scope.save();

    expect(scope.error).toEqual("form.error");
    expect(scope.showError).toBeTruthy();
  });

  it('should insert supply Line', function () {
    scope.supplyLine = {"program": {"name": "HIV"},
      "supplyingFacility": {"code": "F10", "name": "village dispensary"},
      "supervisoryNode": {"code": "F10", "name": "village dispensary"}};
    scope.supplyLineForm = {"$error": {"required": false}};

    $httpBackend.expectPOST('/supplyLines.json', scope.supplyLine).respond(200, {"success": "Saved successfully", "supplyLine": scope.supplyLine});
    scope.save();
    $httpBackend.flush();

    expect(scope.error).toEqual("");
    expect(scope.showError).toBeFalsy();
    expect(scope.$parent.message).toEqual("Saved successfully");
    expect(scope.$parent.supplyLineId).toEqual(supplyLine.id);
  });

  it('should update supply Line', function () {
    scope.supplyLineForm = {"$error": {"required": false}};
    supplyLine.id = 4;
    supplyLine.supplyingFacility = {"code": "F10", "name": "village dispensary"};
    supplyLine.supervisoryNode = {"code": "F10", "name": "village dispensary"};

    $httpBackend.expectPUT('/supplyLines/' + supplyLine.id + '.json', supplyLine).respond(200, {"success": "Saved successfully", "supplyLineId": supplyLine.id});
    scope.save();
    $httpBackend.flush();

    expect(scope.error).toEqual("");
    expect(scope.showError).toBeFalsy();
    expect(scope.$parent.message).toEqual("Saved successfully");
    expect(scope.$parent.supplyLineId).toEqual(supplyLine.id);
  });

});
